### PR TITLE
Debuggable navigation program

### PR DESCRIPTION
**Pull request by schutm**
The Tea_debug.navigationProgram is normally not debuggable (via Tea_debug.xyzProgram). I've added this possibility.

I'm new to ocaml / bucklescript so please comment on any best practices I've missed, so I'm able to fix these.

Unfortunately due to the indentation change it looks like a big commit. However it is actually pretty small:
1. The debug-function changed, in stead of a function which is able to create a program-struct, it returns now a tuple of a init_debug function and debuggle update, view, subscriptions and shutdown functions.
2. I've added a debug_progam and a debug_navigation_program function so the other exposed functions (standardProgram, beginnerProgram, etc.) didn't need to change while still being able to move duplicated functionality into the debug-function and cope with different init-functions.
3. I've added the debug_navigation_progam function.

A diff with whitespace ignored gives a better overview.

Please let me know what to change to get this included.

### CODE REVIEW

**Comment by neochrome** [comment]
This is super-cool! I'll check it out since I've been missing this feature recently and was thinking to add it myself (I implemented the original debugger).

**Comment by neochrome** [comment]
I think there's a small oversight when the common parts where extracted - the `client_msg` function was left behind inside the `debug` function. I think adding the attribute `[@@bs.deriving accessors]` to the debug message type is a ok way forward, please see attached patch.
[fix.diff.txt](https://github.com/OvermindDL1/bucklescript-tea/files/3039133/fix.diff.txt)

**Comment by OvermindDL1** [comment]
> Unfortunately due to the indentation change it looks like a big commit. However it is actually pretty small:

Yeah refmt is a bit irritating and ugly in its formatting...  ^.^;

A more readable github diff (though still not great because refmt):
https://github.com/OvermindDL1/bucklescript-tea/pull/117/files?w=1

Comments inline.  :-)

File: src-ocaml/tea_debug.ml [hunk 1]
```diff
@@ -14,262 +14,304 @@ type 'model debug_model = {
...
-  fun string_of_msg { init; update; view; subscriptions; shutdown } ->
-    let init' (flags : 'flags) =
-      let cmodel, cmd = init flags in
-      {
-        history = ["_init_", cmodel];
-        state = Running;
-        show_details = false;
-      }, cmd |> Tea_cmd.map client_msg
-    in
+let debug string_of_msg update view subscriptions shutdown =
```
Comment (OvermindDL1) [inline src-ocaml/tea_debug.ml]: Any chance of adding types to this function again, when they get complex enough I like to add them for documentative purposes (and to help ensure that I don't break it in the future ^.^; ).  :-)

**Comment by OvermindDL1** [comment]
Overall it looks pretty good, I'll just need to do some testing with it when I get time next.  Please pester me to do so as that is how I prioritize the time I get!  ^.^

**Comment by schutm** [comment]
> I think there's a small oversight when the common parts where extracted - the `client_msg` function was left behind inside the `debug` function. I think adding the attribute `[@@bs.deriving accessors]` to the debug message type is a ok way forward, please see attached patch.

I just copied it verbatim from the original source and kept it as close to it's source location as possible. Personally I like the attribute `[@@bs.deriving.accessors]`. However I think it moves away from standard ocaml. I read somewhere OvermindDL1 was/is thinking about a tea using js_of_ocaml. So I'd like to have his opinion on this one.

**Comment by schutm** [comment]
Accidentally closed. Still have to get used to github...

File: src-ocaml/tea_debug.ml (hunk 1 above)
Comment (schutm) [reply src-ocaml/tea_debug.ml]: Done.

**Comment by OvermindDL1** [comment]
Yeah I'm not a fan of bucklescripts changes the past year or so, it's going away from OCaml and making porting more difficult.... ^.^;

**Comment by neochrome** [comment]
We don't need to use the attribute, but then I think the function
`client_msg` needs to be moved outside the scope of the `debug` function in
order for things to work, or I am missing something?

**Comment by schutm** [comment]
> We don't need to use the attribute, but then I think the function `client_msg` needs to be moved outside the scope of the `debug` function in order for things to work, or I am missing something?

100% correct. Copy/paste error from my private Tea_Extensions module...

**Comment by OvermindDL1** [comment]
I'm leaning to merging this, it looks good and has some good changes.  Anything else you want to do to it first?  :-)

**Comment by schutm** [comment]
Nothing comes to mind at the moment.Best regards,--MartinOp 20 feb. 2020 16:15 schreef OvermindDL1 <notifications@github.com>:I'm leaning to merging this, it looks good and has some good changes.  Anything else you want to do to it first?  :-)

**Comment by OvermindDL1** [comment]
Merged and testing locally, in next release.  :-)

**Pull request merged**